lib/itertools/stream/stream.go

```diff
@@ -0,0 +1,459 @@
+// items when fn returns false.
+func FilterMap[A, B any](stream Stream[A], fn func(A) (B, bool)) Stream[B] {
+	return func(yield func(B, error) bool) {
+		for item, err := range stream {
+			if err != nil {
+				var zero B
+				yield(zero, trace.Wrap(err))
+				return
+			}
+
+			mapped, ok := fn(item)
+			if !ok {
+				continue
+			}
+
+			if !yield(mapped, nil) {
+				return
+			}
+		}
+	}
+}
+
+// MapWhile maps a stream of type A into a stream of type B, halting early
+// if fn returns false.
+func MapWhile[A, B any](stream Stream[A], fn func(A) (B, bool)) Stream[B] {
+	return func(yield func(B, error) bool) {
+		for item, err := range stream {
+			if err != nil {
+				var zero B
+				yield(zero, trace.Wrap(err))
+				return
+			}
+
+			mapped, ok := fn(item)
+			if !ok {
+				return
+			}
+
+			if !yield(mapped, nil) {
+				return
+			}
+		}
+	}
+}
+
+// Chain joins multiple streams in order, fully consuming one before moving to the next.
+func Chain[T any](streams ...Stream[T]) Stream[T] {
+	return func(yield func(T, error) bool) {
+		for _, stream := range streams {
+			for item, err := range stream {
+				if err != nil {
+					var zero T
+					yield(zero, trace.Wrap(err))
+					return
+				}
+
+				if !yield(item, nil) {
+					return
+				}
+			}
+		}
+	}
+}
+
+// Chunks breaks a stream into chunks of a fixed size. The last chunk may be smaller
+// than the specified size. Zero/negative values of size result in an empty stream.
+func Chunks[T any](stream Stream[T], size int) Stream[[]T] {
+	if size < 1 {
+		return Empty[[]T]()
+	}
+	return func(yield func([]T, error) bool) {
+		var chunk []T
+		for item, err := range stream {
+			if err != nil {
+				yield(nil, trace.Wrap(err))
+				return
+			}
+
+			if chunk == nil {
+				chunk = make([]T, 0, size)
+			}
+
+			chunk = append(chunk, item)
+			if len(chunk) == size {
+				if !yield(chunk, nil) {
+					return
+				}
+				chunk = nil
+			}
+		}
+
+		if len(chunk) > 0 {
+			if !yield(chunk, nil) {
+				return
+			}
+		}
+	}
+}
+
+// Fail creates an empty stream that fails immediately with the supplied error.
+func Fail[T any](err error) Stream[T] {
+	return func(yield func(T, error) bool) {
+		if err != nil {
+			var zero T
+			yield(zero, trace.Wrap(err))
+		}
+	}
+}
+
+// Empty creates an empty stream.
+func Empty[T any]() Stream[T] {
+	return func(yield func(T, error) bool) {}
+}
+
+// Once creates a stream that yields a single item.
+func Once[T any](item T) Stream[T] {
+	return func(yield func(T, error) bool) {
+		yield(item, nil)
+	}
+}
+
+// OnceFunc builds a stream from a closure that will yield exactly zero or one items. This stream
+// is the lazy equivalent of the Once/Fail/Empty combinators. A nil error value results
+// in a single-element stream. An error value of io.EOF results in an empty stream. All other error
+// values result in a failing stream.
+func OnceFunc[T any](fn func() (T, error)) Stream[T] {
```

Comment:
I don't like the name because `fn` is not called only once. What's the use for this, btw? It seems like an odd construction.

Reply:
The only current usage is in combination with `Chain(...)` to conditionally add additional values to the end of a stream after the rest of it has already been consumed.

The name collision with `sync.OnceFunc` is unfortunate.  I'm hesitant to change it, since I like the fact that this currently mirrors its equivalent legacy stream combinator.  Open to suggestions though.